src/realm/sync/client_syncsocket.hpp

```diff
@@ -0,0 +1,229 @@
+
+    /// Called whenever a full message has arrived. The WebSocket implementation
+    /// is responsible for defragmenting fragmented messages internally and
+    /// delivering a full message to the Sync Client.
+    ///
+    /// @param data A util::Span containing the data received from the server.
+    ///             The buffer is only valid until the function returns.
+    ///
+    /// @return bool designates whether the WebSocket object should continue
+    ///         processing messages. The normal return value is true . False must
+    ///         be returned if the websocket object has been destroyed during
+    ///         execution of the function.
+    virtual bool websocket_binary_message_received(util::Span<const char> data) = 0;
+
+    /// Called whenever the WebSocket connection has been closed, either as a result
+    /// of a WebSocket error or a normal close.
+    ///
+    /// @param was_clean Was the TCP connection closed after the WebSocket closing
+    ///                  handshake was completed.
+    /// @param status A Status object containing the WebSocket status code and the
+    ///               reason string why the connection was closed.
+    ///
+    /// @return bool designates whether the WebSocket object has been destroyed
+    ///         during the execution of this function. The normal return value is
+    ///         True to indicate the WebSocket object is no longer valid. If False
+    ///         is returned, the WebSocket object will be destroyed at some point
+    ///         in the future.
+    virtual bool websocket_closed_handler(bool was_clean, Status status) = 0;
+};
+
+
+/// Sync Socket interface that provides the event loop and WebSocket factory
+/// used by the SyncClient.
+///
+/// All callback and event operations in the SyncClient must be completed in
+/// the order in which they were issued (via post() or timer) to the event
+/// loop and cannot be run in parallel. It is up to the custom event loop
+/// implementation to determine if these are run on the same thread or a
+/// thread pool as long as it is guaranteed that the callback handler
+/// functions are processed in order and not run concurrently.
+///
+/// The implementation of a SyncSocketInterface must support the following
+/// operations that post handler functions (via by the Sync client) onto the
+/// event loop:
+/// * Post a handler function directly onto the event loop
+/// * Post a handler function when the specified timer duration expires
+///
+/// The event loop is not required to be a single thread as long as the
+/// following requirements are satisfied:
+/// * handler functions are called in the order they were posted to the
+/// event loop queue, and
+/// * a handler function runs to completion before the next handler function
+/// is called.
+///
+/// The SyncSocketInterface also provides a WebSocket interface for
+/// connecting to the server via a WebSocket connection.
+class SyncSocketInterface {
```

Comment:
How about SyncSocketProviderInterface ? SyncSocketInterface kind of implies this _is_ the sync websocket and not just a thing that provides/supports them